File: 238-product-of-array-except-self/238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> leftProds = {};
        vector<int> rightProds(nums.size(), 0);
        
        int cumProd = 1; 
        for(int i = 0 ; i < nums.size(); i++){
            leftProds.push_back(cumProd);
            cumProd *= nums[i];
        }
        
        cumProd = 1; 
        for(int i = nums.size() - 1 ; i >= 0 ; i--){
            rightProds[i] = cumProd;
            cumProd *= nums[i];
        }
        
        vector<int> result(nums.size(), 0);
        for(int i = 0 ; i < nums.size(); i++){
            result[i] = leftProds[i] * rightProds[i];
        }
        
        return result;
        
    }
};

File: 238-product-of-array-except-self/238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        leftProd, rightProd, n = [], [], len(nums)
        prod = 1
        for i in range(n):
            leftProd.append(prod * nums[i])
            prod *= nums[i]
        prod = 1
        for i in range(n - 1, -1, -1):
            rightProd.append(prod * nums[i])
            prod *= nums[i]
        rightProd = rightProd[::-1]
        res = []
        for i in range(n):
            if i == 0:
                res.append(rightProd[1])
            elif i == n - 1:
                res.append(leftProd[n - 2])
            else:
                res.append(rightProd[i + 1] * leftProd[i - 1])
            
        print(leftProd)
        print(rightProd)
        return res